File: src/Z.Data/System.Data.DataRow/DataRow.ToExpandoObject.cs
// Description: C# Extension Methods Library to enhances the .NET Framework by adding hundreds of new methods. It drastically increases developers productivity and code readability. Support C# and VB.NET
// Website & Documentation: https://github.com/zzzprojects/Z.ExtensionMethods
// Forum: https://github.com/zzzprojects/Z.ExtensionMethods/issues
// License: https://github.com/zzzprojects/Z.ExtensionMethods/blob/master/LICENSE
// More projects: http://www.zzzprojects.com/
// Copyright © ZZZ Projects Inc. 2014 - 2016. All rights reserved.
using System.Collections.Generic;
using System.Data;
using System.Dynamic;

public static partial class Extensions
{
    /// <summary>A DataRow extension method that converts the @this to an expando object.</summary>
    /// <param name="this">The @this to act on.</param>
    /// <returns>@this as a dynamic.</returns>
    public static dynamic ToExpandoObject(this DataRow @this)
    {
        dynamic entity = new ExpandoObject();
        var expandoDict = (IDictionary<string, object>) entity;

        foreach (DataColumn column in @this.Table.Columns)
        {
            expandoDict.Add(column.ColumnName, @this[column]);
        }

        return expandoDict;
    }
}

File: src (VB.NET)/Z.ExtensionMethods.VB/Z.Data/System.Data.DataRow/DataRow.ToExpandoObject.vb

' Copyright (c) 2015 ZZZ Projects. All rights reserved
' Licensed under MIT License (MIT) (https://github.com/zzzprojects/Z.ExtensionMethods)
' Website: http://www.zzzprojects.com/
' Feedback / Feature Requests / Issues : http://zzzprojects.uservoice.com/forums/283927
' All ZZZ Projects products: Entity Framework Extensions / Bulk Operations / Extension Methods /Icon Library

Imports System.Collections.Generic
Imports System.Data
Imports System.Dynamic

Public Module Extensions_666

	''' <summary>A DataRow extension method that converts the @this to an expando object.</summary>
	''' <param name="this">The @this to act on.</param>
	''' <returns>@this as a dynamic.</returns>
    <System.Runtime.CompilerServices.Extension> _
    Public Function ToExpandoObject(this As DataRow) As Object
        Dim entity As Object = New ExpandoObject()
        Dim expandoDict = DirectCast(entity, IDictionary(Of String, Object))

        For Each column As DataColumn In this.Table.Columns
            expandoDict.Add(column.ColumnName, this(column))
        Next

        Return expandoDict
    End Function
End Module


